src/lib/provider.ts

```diff
@@ -19,33 +19,54 @@ import {filter} from 'rxjs/operators/filter';
 
 import {EventReplayer} from './api/event.replayer';
 import {PREBOOT_NONCE} from './common/tokens';
-import {getInlinePrebootCode} from './api/inline.preboot.code';
+import {getInlinePrebootCodeDefinition, getInlinePrebootCodeInvocation}
+  from './api/inline.preboot.code';
 import {PrebootOptions} from './common/preboot.interfaces';
+import {validateOptions} from './api';
 
-const PREBOOT_SCRIPT_ID = 'preboot-inline-script';
+const PREBOOT_SCRIPT_CLASS = 'preboot-inline-script';
 export const PREBOOT_OPTIONS = new InjectionToken<PrebootOptions>('PrebootOptions');
 
+function createScriptFromCode(doc: Document, nonce: string|null, inlineCode: string) {
+  const script = doc.createElement('script');
+  if (nonce) {
+    (<any>script)['nonce'] = nonce;
+  }
+  script.className = PREBOOT_SCRIPT_CLASS;
+  script.textContent = inlineCode;
+
+  return script;
+}
+
 export function PREBOOT_FACTORY(doc: Document,
                                 prebootOpts: PrebootOptions,
                                 nonce: string|null,
                                 platformId: Object,
                                 appRef: ApplicationRef,
                                 eventReplayer: EventReplayer) {
   return () => {
+    validateOptions(prebootOpts);
+
     if (isPlatformServer(platformId)) {
-      const inlineCode = getInlinePrebootCode(prebootOpts);
-      const script = doc.createElement('script');
-      if (nonce) {
-        (<any>script)['nonce'] = nonce;
-      }
-      script.id = PREBOOT_SCRIPT_ID;
-      script.textContent = inlineCode;
-      const existingScripts = doc.querySelectorAll(`#${PREBOOT_SCRIPT_ID}`);
+      const inlineCodeDefinition = getInlinePrebootCodeDefinition(prebootOpts);
+      const scriptWithDefinition = createScriptFromCode(doc, nonce, inlineCodeDefinition);
+      const existingScripts = doc.getElementsByClassName(PREBOOT_SCRIPT_CLASS);
 
-      // Check to see if a preboot script is already inlined before adding
-      // it to the DOM. If it is, update the nonce to be current
+      // Check to see if preboot scripts are already inlined before adding them
+      // to the DOM. If they are, update the nonce to be current.
       if (existingScripts.length === 0) {
-        doc.head.appendChild(script);
+        const baseList: string[] = [];
+        const appRootSelectors = baseList.concat(prebootOpts.appRoot);
+        doc.head.appendChild(scriptWithDefinition);
+        appRootSelectors
+          .map(selector => doc.querySelector(selector))
+          .forEach(appRootElem => {
+            if (appRootElem) {
+              const inlineCodeInvocation = getInlinePrebootCodeInvocation(prebootOpts);
+              const scriptWithInvocation = createScriptFromCode(doc, nonce, inlineCodeInvocation);
```

Comment:
In the current version this could be moved out of the loop, near the place where `inlineCodeDefinition` is created. However, since IE doesn't have `document.currentScript` and the [polyfill for that](https://github.com/JamesMGreene/document.currentScript) works only in IE 6-10 we'll need to tag each of those scripts with a unique attribute & inject its value as a variable into the script. Therefore, each of those scripts will be distinct so we'll need to create them in the loop anyway.

The only other solution I see is to trop IE support in Preboot. But I assume that's off the table. ;)

Reply:
Sounds fine